File: services/recommendation-service/src/recommendation_engine.py
from __future__ import annotations

from typing import Any
from recommendation_model import ClassRecommendation, RecommendationInput
from class_rules import recommend as rules_recommend


def generate_recommendation(profile: dict, allowed_evidence: list[dict], campus: str | None = None, role: str = "parent", consultation_stage: str = "NEEDS_ASSESSMENT") -> ClassRecommendation:
    inp = RecommendationInput(profile=profile, allowed_evidence=allowed_evidence, campus=campus, role=role, consultation_stage=consultation_stage)
    rec = rules_recommend(inp)
    
    # Add missing profile fields to missing_info
    known_keys = set(profile.keys())
    needed = ["current_score", "subject_type", "target_school_level", "weak_subjects", "self_discipline_level", "budget_range", "preferred_campus", "boarding_preference"]
    for key in needed:
        if key not in known_keys or not profile.get(key):
            if key not in rec.missing_info:
                rec.missing_info.append(key)
    
    return rec

File: services/recommendation-service/ts-src/recommendationEngine.ts
import {
  type RecommendationInput,
  type ClassRecommendation,
  DEFAULT_RECOMMENDATION_INPUT,
} from "./recommendationModel.js";
import { recommend } from "./classRules.js";

const NEEDED_PROFILE_KEYS = [
  "current_score",
  "subject_type",
  "target_school_level",
  "weak_subjects",
  "self_discipline_level",
  "budget_range",
  "preferred_campus",
  "boarding_preference",
] as const;

export function generateRecommendation(
  profile: Record<string, unknown>,
  allowedEvidence: Record<string, unknown>[],
  campus?: string | null,
  role?: string,
  consultationStage?: string,
): ClassRecommendation {
  const input: RecommendationInput = {
    profile,
    allowed_evidence: allowedEvidence,
    campus: campus ?? DEFAULT_RECOMMENDATION_INPUT.campus,
    role: role ?? DEFAULT_RECOMMENDATION_INPUT.role,
    consultation_stage: consultationStage ?? DEFAULT_RECOMMENDATION_INPUT.consultation_stage,
  };

  const rec = recommend(input.profile, input.allowed_evidence);

  const knownKeys = new Set(Object.keys(profile));
  for (const key of NEEDED_PROFILE_KEYS) {
    if (!knownKeys.has(key) || !profile[key]) {
      if (!rec.missing_info.includes(key)) {
        rec.missing_info.push(key);
      }
    }
  }

  return rec;
}
